internal/providers/nscloud/registry.go

```diff
@@ -107,52 +97,42 @@ func (dk defaultKeychain) Resolve(ctx context.Context, r authn.Resource) (authn.
 		return authn.Anonymous, nil
 	}
 
-	ref, err := name.ParseReference(r.String())
-	if err != nil {
-		return nil, err
-	}
-
-	values := url.Values{}
-	values.Add("scope", fmt.Sprintf("repository:%s:push,pull", ref.Context().RepositoryStr()))
-	values.Add("service", "Authentication")
-
-	req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("https://%s?%s", loginEndpoint, values.Encode()), nil)
-	if err != nil {
-		return nil, err
-	}
+	exchangeToken := func(ctx context.Context) (string, error) {
+		tenantToken, err := auth.LoadTenantToken()
+		if err == nil {
+			resp, err := fnapi.ExchangeTenantToken(ctx, tenantToken.TenantToken, []string{"image-registry-access"})
+			if err != nil {
+				return "", err
+			}
 
-	tok, err := auth.GenerateToken(ctx)
-	if err != nil {
-		return nil, err
-	}
+			return resp.TenantToken, nil
+		}
 
-	req.Header.Add("X-Namespace-Token", tok)
-	req.Header.Add("Authorization", "Bearer "+tok)
+		if !os.IsNotExist(err) {
+			return "", err
+		}
 
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return nil, err
-	}
+		// TODO exchange user auth for tenant tokens.
```

Comment:
Aren't you already doing this with `fnapi.ExchangeUserToken`?

Reply:
Left over from old way. Dropped.